### PR TITLE
[OCRVS-2790] Fix app not loading when offline

### PR DESCRIPTION
**Pull request by rikukissa**
TL;DR
- Fixed the app not loading when browser is offline
- Adds a better support for strict GraphQL query typing

I also fixed one other view (requires update) view being broken when offline. I realised it was because of insufficient typing for a GraphQL query. I've now added a mechanism (`yarn generate-gateway-types`) for generating TS types from GraphQL queries tagged with `gql`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,7 +11,7 @@
     }
   },
   "scripts": {
-    "postinstall": "patch-package && if [ \"$NODE_ENV\" != \"production\" ]; then npx crlf --set=LF ./node_modules/graphql-schema-typescript/**/*; fi",
+    "postinstall": "patch-package",
```
Comment (rikukissa) [inline package.json:14]: I moved the LF fix to `gateway` package as the dependency moved under its dependencies

File: packages/client/src/components/Query.tsx [hunk 2]
```diff
@@ -18,17 +18,15 @@
...
 // eslint-disable-next-line no-restricted-imports
-import { Query as ApolloQuery } from 'react-apollo'
+import { Query as ApolloQuery, QueryProps } from 'react-apollo'
 // eslint-disable-next-line no-restricted-imports
 import * as Sentry from '@sentry/browser'
 
-type Props = ComponentProps<ApolloQuery>
-
-export function Query(props: Props) {
+export function Query<T = any>(props: QueryProps<T>) {
```
Comment (rikukissa) [inline packages/client/src/components/Query.tsx:27]: This allows us to explicitly define types for Apollo queries

Usage:

```ts
<Query<SearchDeclarationsUserWiseQuery>
  query={SEARCH_DECLARATIONS_USER_WISE}
  variables={{
    userId: userDetails!.practitionerId,
  }}
>
  {({ loading, error, data }) => {
```

File: packages/client/src/forms/register/queries/registration.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@ import {
 } from '@opencrvs/gateway/src/graphql/schema.d'
 
 export const FETCH_REGISTRATION = gql`
-  query fetchRegistration($identifier: ID!) {
+  query queryRegistrationByIdentifier($identifier: ID!) {
```
Comment (rikukissa) [inline packages/client/src/forms/register/queries/registration.ts:20]: I had to add a name / rename almost all of our queries so the type generator is able to use unique names for the new types. The result type for this query looks like this

```ts
export type QueryRegistrationByIdentifierQuery = {
  __typename?: 'Query'
  queryRegistrationByIdentifier?: {
    __typename?: 'BirthRegistration'
    id: string
    child?: {
      __typename?: 'Person'
      id?: string | null
      birthDate?: string | null
      gender?: string | null
      name?: Array<{
        __typename?: 'HumanName'
        use?: string | null
        firstNames?: string | null
        familyName?: string | null
      } | null> | null
    } | null
  } | null
}
```

File: packages/client/src/src-sw.js [hunk 4]
```diff
@@ -71,7 +71,7 @@ workbox.precaching.precacheAndRoute([])
  */
 
 workbox.routing.registerRoute(
-  '/config.js',
+  /http(.+)config\.js$/,
```
Comment (rikukissa) [inline packages/client/src/src-sw.js:74]: Not the best regexp, I know. There was a tiny undocumented constraint with the `registerRoute` – it only lets through cross-domain URLs if the regexp given matches the url from the beginning, so from position 0.

https://github.com/GoogleChrome/workbox/blob/0cc6975f17b60d67a71d8b717e73ef7ddb79a891/packages/workbox-routing/src/RegExpRoute.ts#L69

Comment (rikukissa) [reply packages/client/src/src-sw.js:74]: I think this is something that prevented the app from working in offline completely as it is now

File: packages/client/src/utils/gateway.ts [hunk 5]
```diff
@@ -0,0 +1,4255 @@
...
+  id?: Maybe<Scalars['String']>
+  mobile?: Maybe<Scalars['String']>
+  scrope?: Maybe<Array<Maybe<Scalars['String']>>>
+  status?: Maybe<Scalars['String']>
+  username?: Maybe<Scalars['String']>
+}
+
+export type Unnamed_1_QueryVariables = Exact<{ [key: string]: never }>
+
+export type Unnamed_1_Query = {
```
Comment (Zangetsu101) [inline packages/client/src/utils/gateway.ts:1695]: You must have missed this one

File: packages/client/src/utils/gateway.ts [hunk 6]
```diff
@@ -0,0 +1,4255 @@
...
+        firstNames?: string | null
+        familyName?: string | null
+      } | null> | null
+    } | null
+  } | null
+}
+
+export type Unnamed_2_QueryVariables = Exact<{ [key: string]: never }>
+
+export type Unnamed_2_Query = {
```
Comment (Zangetsu101) [inline packages/client/src/utils/gateway.ts:1786]: This one too

File: packages/client/src/views/OfficeHome/queries.ts [hunk 7]
```diff
@@ -11,39 +11,6 @@
...
-    ... on DeathEventSearchSet {
-      dateOfDeath
-      deceasedName {
-        firstNames
-        familyName
-        use
-      }
-    }
-  `
-
```
Comment (Zangetsu101) [inline packages/client/src/views/OfficeHome/queries.ts:46]: Was there any particular reason for removing this? Maybe we should use [Fragments](https://www.apollographql.com/docs/react/data/fragments/#example-usage) instead?

**Review by Zangetsu101 — CHANGES_REQUESTED** [review]
This is really nice. But there are some queries or mutations which were defined in the file it was used so I guess we won't be getting types generated for those. Looking through all the files will be time consuming I guess.

Also when (if) we move to using `useQuery` and `useMutation` we should be able to use codegen for generating types for those as well

File: packages/client/src/src-sw.js (hunk 4 above)
Comment (euanmillar) [reply packages/client/src/src-sw.js:74]: Amazing!

File: packages/client/src/views/OfficeHome/queries.ts (hunk 7 above)
Comment (rikukissa) [reply packages/client/src/views/OfficeHome/queries.ts:46]: Yea, the AST parser of codegen doesn't support that unfortunately and agreed. Fragments might just work. I'll give it a try

**Comment by rikukissa** [comment]
I'll change the codegen file glob so that it finds all ts(x) files, so 

```
 - 'src/**/*.ts*'
- '!src/utils/gateway.ts'
- '!src/**/*.test.ts*'
 ```
 
 then we should be able to also have queries in components

File: packages/client/src/views/OfficeHome/queries.ts (hunk 7 above)
Comment (rikukissa) [reply packages/client/src/views/OfficeHome/queries.ts:46]: It worked! Thanks for the suggestion

Comment (Zangetsu101) [reply packages/client/src/views/OfficeHome/queries.ts:46]: Ok cool

**Review by Zangetsu101 — APPROVED** [review]

**Pull request merged**